keps/sig-scheduling/4328-node-affinity-based-eviction/README.md

```diff
@@ -0,0 +1,426 @@
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [x] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [x] (R) KEP approvers have approved the KEP status as `implementable`
+- [x] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP aims to introduce node-affinity-based eviction.
+In this KEP, we will introduce a new nodeAffinity type: `requiredDuringSchedulingRequiredDuringExecution`.
+A new controller called `node-affinity-eviction` will be added to the controller-manager. 
+This controller will monitor changes in node labels and evict pods when their `requiredDuringSchedulingRequiredDuringExecution` node affinity is no longer met.
+
+## Motivation
+
+- Enable `requiredDuringSchedulingRequiredDuringExecution` node affinity for pod scheduling and execution.
+- Provide users with greater control over pod placement.
+- Ensure predictable and consistent pod behavior in various scenarios.
+- Address use cases with specific placement and execution requirements.
+  
+### Goals
+
+- Introduce the `RequiredDuringSchedulingRequiredDuringExecution` **nodeAffinity** type.
+- Add `node-affinity-eviction` controller to ensure pods being evicted if the selectors are no longer met at runtime.
+
+### Non-Goals
+
+- This KEP does not aim to introduce the `RequiredDuringSchedulingRequiredDuringExecution` **podAffinity** type and **podAntiAffinity** type.
+- This feature does not respect PodDisruptionBudget during the eviction.
```

Comment:
Pod-affinity-eviction will be more complicated than node-affinity-eviction. We will consider it in the future, but even if we decide to implement it, it will definitely need another KEP.

Reply:
What I refer to is the PDB. Can we consider evictionAPI directly here? @alculquicondor  I know we use Delete op directly in scheduler right now, the taint-eviction, preemption, but evictionAPI seems a better choice, no longer to consider the PDB.

Reply:
> it's hard to respect PDB under this framework.

doesn't the evictionAPI handle that for you?

TBH, I'm not really sure which approach to take, and I'm not sure why taint based evictions chose not to respect PDB. It might be that both features were developed in parallel without paying much attention to each other.

Let me try to pull in some long timers: @damemi @bsalamat @liggitt @wojtek-t

Reply:
> doesn't the evictionAPI handle that for you?

What I concerned is how we handle the evict failure due to PDB.
If we choose to save the pod info in memory and retry the evict every once in a while, It will bring a lot of memory consumption and CPU consumption in a large cluster.

So I think we need to change the current controller framework(listening to events and trigger deletion) to a framework similar to deschduler (there is a loop to continuously check each pod) to reduce memory consumption, but it may increase CPU consumption.

Reply:
This is a problem either way, delete(with PDB) or evict. I would like the lightweight approach - ignore the evict error for PDB, and people can choose rescheduler for a strict policy. Seems we found the boundary... I don't know.

Reply:
@kerthcet @alculquicondor Just came up with another thought: Why don't we let users to decide?

So I found this KEP: https://github.com/kubernetes/enhancements/issues/3280
It seems that we introduced a new field `AllowDisruptionByPriorityGreaterThanOrEqual`, which can allow users to decide whether this pod can be evicted when eviction will violate PDBs.

We can do the same for this node-affinity, maybe sth like this:
```
type RequiredExecutionTerm struct {
	Required          NodeSelectorTerm
	TolerationSeconds *int64
        RespectPDB        bool
}
```

If `RespectPDB ` is set to true, we evict pods with evictionAPI and ignore the evict error.
If not, we just delete the pod.

Reply:
I would start as simple as possible and just respect PDB in all scenarios (no boolean and no tolerationSeconds).

Is there any reason (other than implementation details) that this wouldn't make sense for your stories? It fits the migration story in the case of multitenant nodes.

Reply:
If it's just for `migration story`, `just respect PDB in all scenarios (no boolean and no tolerationSeconds)` this is totally ok.

Reply:
So far we have seen problems with every kind of an eviction/deletion and I think it would be better to be cautious here as it is quite hard/impossible to backtrack on these decisions.

There is a potential to hurt the availability of existing workloads. The bright side is that this is configured by the application or a user/entity deploying it. The not-so-bright side is that the app does not have all the information about when and where it will run.
- Is the app rollout in progress?
- Is there a PDB configured? If so, is it blocking?
- Are there enough resources on other nodes once the app is evicted? Is the Kubelet going to repel the Pod?
- Does a migration need to occur, before the Pod should be terminated? Is the entity doing the migration the same entity that is deploying the application?

I am also concerned about what will happen if we make a convention about the behavior and then we add other descheduling effects like the pod affinity. This will increase the changes that the user will shoot themselves in the foot.

I am exploring a new feature in the [Declarative Node Maintenance KEP](https://github.com/kubernetes/enhancements/pull/4213) called `EvacuationRequest`. Which is softer than evictions are and gives more control to the application's hands.


My main point is, is there a reason why the eviction should be immediate? 

Hard to Soft evictions:

`Delete with grace period 0 >  Node Eviction > Delete > Peemption Eviction > Eviction API > Eviction API with PDB > EvacuationRequest`


In my opinion, it would be better to start with the softest approach as the default behavior and if there are use cases that we deem reasonable we could offer configurability of this scale.